### PR TITLE
Parse error recovery and completion for `A::`

### PR DESCRIPTION
**Pull request by jez**
### Motivation

Better completion.

### Test plan

See included automated tests.

### CODE REVIEW

File: parser/Builder.cc
Comment (froydnj) [inline parser/Builder.cc:32]: Why is this needed?

File: main/lsp/requests/completion.cc
Comment (froydnj) [inline main/lsp/requests/completion.cc:995]: I think the rationale in the comment or the logic for `suggestKeywords` needs to be adjusted:

https://github.com/sorbet/sorbet/blob/10481e1af8df9913da6e57e5960fee5c736cbf2e/ast/desugar/Desugar.cc#L645-L648

since we don't want `self.de` to complete with keywords.

Also `super` can be `isPrivateOk`, so you may want to check for that specifically too?

File: main/lsp/requests/completion.cc
Comment (froydnj) [inline main/lsp/requests/completion.cc:1005]: Same comment applies to this use of `isPrivateOk`, I think?

**Review by froydnj — CHANGES_REQUESTED** [review]
It looks like there are some test failures.  The idea seems reasonable; a few comments below.

File: parser/Builder.cc
Comment (jez) [reply parser/Builder.cc:32]: good catch, this was the result of a bad `git add -p`, it's not needed.
```suggestion
```

File: parser/parser/cc/grammars/typedruby.ypp
Comment (jez) [reply parser/parser/cc/grammars/typedruby.ypp:1723]: sure i can factor a helper. do you mind if i do it in another PR, once everything lands?

File: main/lsp/requests/completion.cc
Comment (jez) [reply main/lsp/requests/completion.cc:995]: Thanks. The first one is an existing bug. I'll open a ticket for it, but I don't want to block this PR on fixing it.

https://github.com/sorbet/sorbet/issues/5157

The `super` question is actually subtle, because it gets translated to `<call-with-block>(<self>, :<super>, <blk>)` which then suggests no completion results. So it does it in a roundabout way but it works.

File: main/lsp/requests/completion.cc
Comment (jez) [reply main/lsp/requests/completion.cc:1005]: https://github.com/sorbet/sorbet/issues/5158

File: parser/parser/cc/grammars/typedruby.ypp
Comment (jez) [reply parser/parser/cc/grammars/typedruby.ypp:1723]: i've made a ticket for this internally

**Review by froydnj — APPROVED** [review]
Thanks for filing the follow-on tickets.  This looks good, but I think there are still some test failures?

**Pull request merged**
